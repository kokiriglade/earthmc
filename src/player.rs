//! # Player
//!
//! Defines the [`Player`] struct.
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{
    named_id::{NamedId, NamedIdOpt},
    nation::NationRankKind,
    permission::TownyPermissions,
    town::TownRankKind,
};

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Player {
    /// The player's name.
    pub name: String,
    /// The player's [`Uuid`].
    pub uuid: Uuid,
    /// The player's title set through `/n set title` in-game.
    pub title: Option<String>,
    /// The player's title set through `/n set surname` in-game.
    pub surname: Option<String>,
    /// Formatted name combining, in this order, title, username and surname.
    pub formatted_name: String,
    /// About section of `/res` set with `/res set about` in-game.
    pub about: Option<String>,
    /// The player's town.
    pub town: NamedIdOpt,
    /// The nation that the player's town is a part of.
    pub nation: NamedIdOpt,
    pub timestamps: PlayerTimestamps,
    pub status: PlayerStatus,
    pub stats: PlayerStats,
    pub perms: TownyPermissions,
    /// The ranks that the player has in both their town and nation.
    pub ranks: PlayerRanks,
    /// List of the player's friends.
    pub friends: Vec<NamedId>,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct PlayerTimestamps {
    /// Unix timestamp representing when the player joined the server.
    pub registered: i64,
    /// Unix timestamp representing when the player joined their town.
    pub joined_town_at: Option<i64>,
    /// Unix timestamp representing when the player was last online.
    /// Can be [`None`] if the player is an NPC.
    pub last_online: Option<i64>,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct PlayerStatus {
    /// If the player is currently online.
    pub is_online: bool,
    /// If the player is a Towny NPC.
    #[serde(rename = "isNPC")]
    pub is_npc: bool,
    /// If the player is a mayor of a town.
    pub is_mayor: bool,
    /// If the player is a leader of a nation.
    pub is_king: bool,
    /// If the player is currently in a town.
    pub has_town: bool,
    /// If the player is currently in a nation.
    pub has_nation: bool,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct PlayerStats {
    /// The player's current balance as seen on `/res` in-game.
    pub balance: f64,
    /// The amount of friends this player has.
    pub num_friends: i32,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct PlayerRanks {
    /// A list of town ranks the player holds.
    pub town_ranks: Vec<TownRankKind>,
    /// A list of nation ranks the player holds.
    pub nation_ranks: Vec<NationRankKind>,
}
