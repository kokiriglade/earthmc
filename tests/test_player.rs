use earthmc::player::Player;

#[test]
fn test_player() {
    let raw_json = include_str!("inputs/player.json");
    let parsed: Vec<Player> = serde_json::from_str(raw_json).unwrap();
    insta::assert_debug_snapshot!(parsed);
}
