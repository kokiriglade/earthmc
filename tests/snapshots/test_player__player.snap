---
source: tests/test_player.rs
expression: parsed
---
[
    Player {
        name: "Fruitloopins",
        uuid: fed0ec4a-f1ad-4b97-9443-876391668b34,
        title: Some(
            "<rainbow>King",
        ),
        surname: Some(
            "of Yukon",
        ),
        formatted_name: "<rainbow>King Fruitloopins of Yukon",
        about: Some(
            "CEO of this feature",
        ),
        town: NamedIdGeneric {
            name: Some(
                "Teslin",
            ),
            uuid: Some(
                969d4960-f082-4be0-9e32-f14ed0b7cd50,
            ),
        },
        nation: NamedIdGeneric {
            name: Some(
                "Yukon",
            ),
            uuid: Some(
                3ce63924-44f9-404a-93e0-2fe87a22edd9,
            ),
        },
        timestamps: PlayerTimestamps {
            registered: 1651299304371,
            joined_town_at: Some(
                1701704799323,
            ),
            last_online: Some(
                1711116371655,
            ),
        },
        status: PlayerStatus {
            is_online: true,
            is_npc: false,
            is_mayor: false,
            is_king: false,
            has_town: true,
            has_nation: true,
        },
        stats: PlayerStats {
            balance: 2838.0,
            num_friends: 1,
        },
        perms: TownyPermissions {
            build: TownyPermissionSet {
                resident: false,
                nation: false,
                ally: false,
                outsider: false,
            },
            destroy: TownyPermissionSet {
                resident: false,
                nation: false,
                ally: false,
                outsider: false,
            },
            switch: TownyPermissionSet {
                resident: false,
                nation: false,
                ally: false,
                outsider: false,
            },
            item_use: TownyPermissionSet {
                resident: false,
                nation: false,
                ally: false,
                outsider: false,
            },
            flags: TownyPermissionFlags {
                pvp: false,
                explosion: false,
                fire: false,
                mobs: false,
            },
        },
        ranks: PlayerRanks {
            town_ranks: [
                Councillor,
            ],
            nation_ranks: [
                Chancellor,
            ],
        },
        friends: [
            NamedIdGeneric {
                name: "tuzzzie",
                uuid: 8391474f-4b57-412a-a835-96bd2c253219,
            },
        ],
    },
]
